### PR TITLE
[Merged by Bors] - feat : define CardinalInterFilter

### PR DESCRIPTION
**Pull request by JADekker**
CardinalInterFilter generalizes CountableInterFilter to allow for a specific cardinality, where the filter must be stable under intersections over a number of sets lower than this cardinality.

Prepares the way for K-Lindelöf spaces.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by JADekker** [comment]
The current PR is deliberately kept small: once the exact formulation has been agreed upon, I will mimic the other results in CountableInterFilter.lean and then derive the results there as a special case of those of CardinalInterFilter

**Review by ocfnash — COMMENTED** [review]
Thank you for doing this; many apologies for being so slow to review.

I was going to suggest that we do not need to have both `CardinalInterFilter` and `CountableInterFilter` and that instead of adding this new definition could we just replace `CountableInterFilter` with `CardinalInterFilter` and:
```
abbrev CountableInterFilter (l : Filter α) := CardinalInterFilter l (aleph 1)
```
as well as API for `CountableInterFilter` to connect with `Set.Countable`.

However it is not clear to me that every user of `CountableInterFilter` should need to import the full theory of ordinals, so this seems like a reasonable case for having two definitions (at least for now).

The code looks great, I have some nitpicks below and I'll take another pass quickly after you have replied.

File: Mathlib/Order/Filter/CardinalInter.lean [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+/-- Every CardinalInterFilter for some c also is a CardinalInterFilter for some a < c -/
+theorem CardinalInterFilter.to_lower_cardinality (l : Filter α) [CardinalInterFilter l c]
+    {a : Cardinal.{u}} (hac : a < c) :
+  CardinalInterFilter l a where
+    cardinal_sInter_mem :=
+      fun S hS a ↦ CardinalInterFilter.cardinal_sInter_mem S (lt_trans hS hac) a
+
+-- NOTE: Some basic API to show that this works, initial reviews may concentrate on the above
+-- definitions first.
```
[the remark this answers is not in the document]
Comment (JADekker) [reply Mathlib/Order/Filter/CardinalInter.lean:81]: I propose we postpone that until we know if we will refactor `CountableInterFilter` in terms of `CardinalInterFilter`, which is something I want to PR after this PR merges.

**Review by ocfnash — APPROVED** [review]
Thanks again for doing this.

Please also add a comment to the `CountableInterFilter` file saying something like:
```
Note that there also exists a typeclass `CardinalInterFilter`, and thus an alternative spelling of
`CountableInterFilter` as `CardinalInterFilter l (aleph 1)`. The former (defined here) is the
preferred spelling; it has the advantage of not requiring the user to import the theory ordinals.
```
(or similar)

bors d+

**Pull request closed without merging**